File: leetcode/binary_tree_right_side_view/binary_tree_right_side_view.java
package binary_tree_right_side_view;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if(root == null) return ans;

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        int level = 1;
        TreeNode curr;

        while(!q.isEmpty()) {
            curr = q.poll();

            if(curr.left != null) q.add(curr.left);
            if(curr.right != null) q.add(curr.right);

            if(--level == 0) {
                ans.add(curr.val);
                level = q.size();
            }
        }
        return ans;
    }
}

File: leetcode/binary_tree_right_side_view/binary_tree_right_side_view.cpp
/*
 * Time Complexity: O(N)
 *   - N is the number of nodes in the tree.
 *   - Each node is visited once during the level order traversal.
 *
 * Space Complexity: O(W)
 *   - W is the maximum width of the tree (maximum number of nodes at any level).
 *   - Due to the queue used for level order traversal.
 */

#include <vector>
#include <queue>

using std::vector;
using std::queue;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans;
        if(root == nullptr) return ans;
    
        queue<TreeNode*> q;
        q.push(root);
        int level = 1;
        TreeNode* curr;

        while(!q.empty()) {
            curr = q.front();
            q.pop();
            
            if(curr->left) q.push(curr->left);
            if(curr->right) q.push(curr->right);

            if(--level == 0) {
                ans.push_back(curr->val);
                level = q.size();
            }
        }

        return ans;
    }
};